go-controller/pkg/ovn/namespace.go

```diff
@@ -67,11 +70,38 @@ func (oc *DefaultNetworkController) syncNamespaces(namespaces []interface{}) err
 		expectedNs[ns.Name] = true
 	}
 
-	err := oc.addressSetFactory.ProcessEachAddressSet(func(addrSetName, namespaceName, nameSuffix string) error {
-		if nameSuffix == "" && !expectedNs[namespaceName] {
-			if err := oc.addressSetFactory.DestroyAddressSetInBackingStore(addrSetName); err != nil {
-				klog.Errorf(err.Error())
-				return err
+	err := oc.addressSetFactory.ProcessEachAddressSet(func(hashedName, addrSetName string) error {
+		// filter out address sets owned by HybridRoutePolicy and EgressQoS by prefix, and
+		// owned by network policy by dot in the name (namespace can't have dots in its name).
+		// the only left address sets may be owned by egress firewall dns or namespace
+		if !strings.HasPrefix(addrSetName, types.HybridRoutePolicyPrefix) &&
+			!strings.HasPrefix(addrSetName, types.EgressQoSRulePrefix) &&
+			!strings.Contains(addrSetName, ".") {
+			// make sure address set is not owned by egress firewall dns
+			// find ACLs referencing given address set (by hashName)
+			aclPred := func(acl *nbdb.ACL) bool {
+				return strings.Contains(acl.Match, "$"+hashedName)
+			}
+			acls, err := libovsdbops.FindACLsWithPredicate(oc.nbClient, aclPred)
+			if err != nil {
+				return fmt.Errorf("failed to find referencing acls for address set %s: %v", addrSetName, err)
+			}
+			if len(acls) > 0 {
+				// if given address set is owned by egress firewall, all ACLs will be owned by the same object
+				acl := acls[0]
+				// check if egress firewall dns is the owner
+				// the only address set that may be referenced in egress firewall destination is dns address set
+				if acl.ExternalIDs[egressFirewallACLExtIdKey] != "" && strings.Contains(acl.Match, ".dst == $"+hashedName) {
+					// address set is owned by egress firewall, skip
+					return nil
+				}
+			}
+			// address set is owned by namespace, namespace name = address set name
+			if !expectedNs[addrSetName] {
```

Comment:
This part is a little weird to me. Regardless of `FindACLsWithPredicate` wouldn't we want to `oc.addressSetFactory.DestroyAddressSetInBackingStore` if a corresponding namespace was not found?